### ISSUE TITLE
Issues with queries containing OR and multiple nodes in the disjuncts

### ISSUE THREAD

**Issue by fab-bar** [korpling/ANNIS#584]
I have stumbled over issues with queries containing an OR (|) and multiple nodes in the disjuncts with frequency analysis and csv-Export. All of these issues exist in version 3.5.0-preview7.

Example:

    (zu#pos=‎"PTKZU‎" . vb#pos=‎/V.INF‎/)|(zu#pos=‎"APPR‎" . vb#pos=‎"NA‎")

(https://annis.corpora.uni-hamburg.de/gui/?id=d8424564-1c22-4ff3-bdaa-b8d3afbe6e46)

Works as expected in the frequency analysis and the csv-Export while for the following query the the wrong nodes are paired:

    (zu#pos=‎"PTKZU‎" . vb#pos=‎/V.INF‎/)|(vb#pos=‎"NA‎" & zu#pos=‎"APPR‎" & #zu . #vb)

(https://annis.corpora.uni-hamburg.de/gui/?id=270c0692-c116-4103-bcee-d111151e3b44)

#zu from the first disjunct is paired with #vb from the second disjunct for the frequency analysis, csv-Export and regarding the colors in the AQL query and the results tab.

![Frequency Analysis](https://user-images.githubusercontent.com/11232385/43843702-6fd16bc8-9b28-11e8-9ca7-b9ec15e3be12.png)

It gets more complicated, when the disjuncts contain different numbers of nodes as in the next example:

    (zu#pos=‎"PTKZU‎" . vb#pos=‎/V.INF‎/)|(zu#pos=‎"APPR‎" _=_ lemma_wsd=‎"tô¹‎" . vb#pos=‎"NA‎" _=_ posLemma=‎"VVINF‎")

(https://annis.corpora.uni-hamburg.de/gui/?id=b6d062a8-13f6-4cba-af29-73faf19abfa0)

Here the frequency analysis leads to an error, the csv-Export pairs #vb from the first disjunct with "lemma_wsd" from the second and so on.

Again, the order of the nodes is relevant as can be seen with the following query:

    (zu#pos=‎"PTKZU‎" . vb#pos=‎/V.INF‎/)|(zu#pos=‎"APPR‎" . vb#pos=‎"NA‎" & lemma_wsd=‎"tô¹‎" & posLemma=‎"VVINF‎" & #zu _=_ #5 & #vb _=_ #6)

(https://annis.corpora.uni-hamburg.de/gui/?id=67aa6b4d-4da5-4fb3-b5ac-9a8c730a007d)
    
Here the frequency analysis loads but has empty values for #zu and #vb from the first disjunct.

![Frequency Analysis](https://user-images.githubusercontent.com/11232385/43844141-54068dd2-9b29-11e8-87d7-74e8e24b11f9.png)

To summarize, here is a list of the problems that I identified:

- [x] frequency analysis depends on order of the nodes in query instead of the
      name of the nodes, might even fail
- [x] csv export combines nodes from the matches by the order of the nodes in the disjuncts instead of the name of the nodes 
- [ ] nodes in AQL query with the same name should have the same color
- [ ] the same for matches

I have traced the problem with the frequency analysis and the csv export to the way the SQL-queries are created. The first problem is that the code that creates the "SELECT"s for the disjuncts counts the nodes for each disjunct separately (https://github.com/korpling/ANNIS/blob/4de11bc414e8dc1e4dd236c321444448cf0a1f99/annis-service/src/main/java/annis/sqlgen/AbstractUnionSqlGenerator.java#L55 and https://github.com/korpling/ANNIS/blob/4de11bc414e8dc1e4dd236c321444448cf0a1f99/annis-service/src/main/java/annis/sqlgen/SolutionSqlGenerator.java#L69) and therefore creates "id1" to "id4" (with padding for the first disjunct) in the order that the nodes appear in the query for "solution". In the end, nodes with the same number appear in the same column of the result set. The code that creates the "LEFT JOIN" takes the ids from the nodes for the id references. When doing the frequency analysis for #zu and #vb from

    (zu#pos=‎"PTKZU‎" . vb#pos=‎/V.INF‎/)|(zu#pos=‎"APPR‎" _=_ lemma_wsd=‎"tô¹‎" . vb#pos=‎"NA‎" _=_ posLemma=‎"VVINF‎")

the nodes from the second disjunct are used leading to "id3" and "id5" leading to an exception as only "id1" to "id4" exist in solution.

I am working on some code changes that will address the issues with the frequency analysis.

**Comment by fab-bar** [comment]
I have adapted the code that creates the SQL for queries and the frequency analysis such that named nodes are put to the beginning of "solutions" in alphabetical order of their names with the id containing the name instead of a number. All other nodes are added in the order of their appearance: https://github.com/fab-bar/ANNIS/commit/aaaed109a34468e289ac831cea052ed29aac3827

As far as I have seen, these changes fix the frequency analysis and only break the CSV export - I will look into adapting the MatrixSqlGenerator to fix this as well and prepare a pull request.

Here is one example for the generated SQL query to do the frequency analysis for the AQL query

    (zu#pos=‎"PTKZU‎" . vb#pos=‎/V.INF‎/)|(zu#pos=‎"APPR‎" _=_ lemma_wsd=‎"tô¹‎" . vb#pos=‎"NA‎" _=_ posLemma=‎"VVINF‎")

Originally it is:

```
SELECT v1.node_qannotext AS value1, v2.node_qannotext AS value2, count(*) AS "count"
FROM
  (
  SELECT 
      facts1.id AS id1,
      facts1.node_anno_category AS cat1,
      facts1.salt_id AS salt_id1,
      facts2.id AS id2,
      facts2.node_anno_category AS cat2,
      facts2.salt_id AS salt_id2,
      NULL::bigint AS id3,
      NULL::integer AS cat3,
      NULL::varchar AS salt_id3,
      NULL::bigint AS id4,
      NULL::integer AS cat4,
      NULL::varchar AS salt_id4,
      (SELECT c.path_name FROM corpus AS c WHERE c.id = facts1.corpus_ref LIMIT 1) AS path_name,
      facts1.toplevel_corpus AS toplevel_corpus,
      facts1.corpus_ref AS corpus_ref
  FROM
    facts_442 AS facts1,
    facts_442 AS facts2
  WHERE
    facts1.corpus_ref = facts2.corpus_ref AND
    facts1.n_na_sample IS TRUE AND
    facts1.node_annotext LIKE 'pos:PTKZU' AND
    facts1.right_token = facts2.left_token - 1 AND
    facts1.text_ref = facts2.text_ref AND
    facts1.toplevel_corpus IN (442) AND
    facts2.n_na_sample IS TRUE AND
    facts2.node_annotext ~ '^(pos:(V.INF))$' AND
    facts2.toplevel_corpus IN (442)

  UNION   SELECT 
      facts3.id AS id1,
      facts3.node_anno_category AS cat1,
      facts3.salt_id AS salt_id1,
      facts4.id AS id2,
      facts4.node_anno_category AS cat2,
      facts4.salt_id AS salt_id2,
      facts5.id AS id3,
      facts5.node_anno_category AS cat3,
      facts5.salt_id AS salt_id3,
      facts6.id AS id4,
      facts6.node_anno_category AS cat4,
      facts6.salt_id AS salt_id4,
      (SELECT c.path_name FROM corpus AS c WHERE c.id = facts3.corpus_ref LIMIT 1) AS path_name,
      facts3.toplevel_corpus AS toplevel_corpus,
      facts3.corpus_ref AS corpus_ref
  FROM
    facts_442 AS facts3,
    facts_442 AS facts4,
    facts_442 AS facts5,
    facts_442 AS facts6
  WHERE
    (facts3.id <> facts4.id OR facts3.node_anno_category IS DISTINCT FROM facts4.node_anno_category) AND
    (facts5.id <> facts6.id OR facts5.node_anno_category IS DISTINCT FROM facts6.node_anno_category) AND
    facts3.corpus_ref = facts4.corpus_ref AND
    facts3.left_token = facts4.left_token AND
    facts3.n_na_sample IS TRUE AND
    facts3.node_annotext LIKE 'pos:APPR' AND
    facts3.right_token ^=^ facts4.right_token AND
    facts3.text_ref = facts4.text_ref AND
    facts3.toplevel_corpus IN (442) AND
    facts4.corpus_ref = facts5.corpus_ref AND
    facts4.n_na_sample IS TRUE AND
    facts4.node_annotext LIKE 'lemma\_wsd:tô¹' AND
    facts4.right_token = facts5.left_token - 1 AND
    facts4.text_ref = facts5.text_ref AND
    facts4.toplevel_corpus IN (442) AND
    facts5.corpus_ref = facts6.corpus_ref AND
    facts5.left_token = facts6.left_token AND
    facts5.n_na_sample IS TRUE AND
    facts5.node_annotext LIKE 'pos:NA' AND
    facts5.right_token ^=^ facts6.right_token AND
    facts5.text_ref = facts6.text_ref AND
    facts5.toplevel_corpus IN (442) AND
    facts6.n_na_sample IS TRUE AND
    facts6.node_annotext LIKE 'posLemma:VVINF' AND
    facts6.toplevel_corpus IN (442)

) AS solutions
  LEFT JOIN facts_442 AS v1 ON (v1.toplevel_corpus IN (442) AND v1.toplevel_corpus = solutions.toplevel_corpus AND v1.id = solutions.id5 AND v1.node_annotext LIKE 'pos:%' AND v1.n_na_sample IS TRUE)
  LEFT JOIN facts_442 AS v2 ON (v2.toplevel_corpus IN (442) AND v2.toplevel_corpus = solutions.toplevel_corpus AND v2.id = solutions.id3 AND v2.node_annotext LIKE 'pos:%' AND v2.n_na_sample IS TRUE)

GROUP BY value1, value2

```
With the commited code it is changed to:

```
SELECT v1.node_qannotext AS value1, v2.node_qannotext AS value2, count(*) AS "count"
FROM
  (
  SELECT 
      facts2.id AS idvb,
      facts2.node_anno_category AS catvb,
      facts2.salt_id AS salt_idvb,
      facts1.id AS idzu,
      facts1.node_anno_category AS catzu,
      facts1.salt_id AS salt_idzu,
      NULL::bigint AS id1,
      NULL::integer AS cat1,
      NULL::varchar AS salt_id1,
      NULL::bigint AS id2,
      NULL::integer AS cat2,
      NULL::varchar AS salt_id2,
      (SELECT c.path_name FROM corpus AS c WHERE c.id = facts1.corpus_ref LIMIT 1) AS path_name,
      facts1.toplevel_corpus AS toplevel_corpus,
      facts1.corpus_ref AS corpus_ref
  FROM
    facts_442 AS facts1,
    facts_442 AS facts2
  WHERE
    facts1.corpus_ref = facts2.corpus_ref AND
    facts1.n_na_sample IS TRUE AND
    facts1.node_annotext LIKE 'pos:PTKZU' AND
    facts1.right_token = facts2.left_token - 1 AND
    facts1.text_ref = facts2.text_ref AND
    facts1.toplevel_corpus IN (442) AND
    facts2.n_na_sample IS TRUE AND
    facts2.node_annotext ~ '^(pos:(V.INF))$' AND
    facts2.toplevel_corpus IN (442)

  UNION   SELECT 
      facts5.id AS idvb,
      facts5.node_anno_category AS catvb,
      facts5.salt_id AS salt_idvb,
      facts3.id AS idzu,
      facts3.node_anno_category AS catzu,
      facts3.salt_id AS salt_idzu,
      facts4.id AS id1,
      facts4.node_anno_category AS cat1,
      facts4.salt_id AS salt_id1,
      facts6.id AS id2,
      facts6.node_anno_category AS cat2,
      facts6.salt_id AS salt_id2,
      (SELECT c.path_name FROM corpus AS c WHERE c.id = facts3.corpus_ref LIMIT 1) AS path_name,
      facts3.toplevel_corpus AS toplevel_corpus,
      facts3.corpus_ref AS corpus_ref
  FROM
    facts_442 AS facts3,
    facts_442 AS facts4,
    facts_442 AS facts5,
    facts_442 AS facts6
  WHERE
    (facts3.id <> facts4.id OR facts3.node_anno_category IS DISTINCT FROM facts4.node_anno_category) AND
    (facts5.id <> facts6.id OR facts5.node_anno_category IS DISTINCT FROM facts6.node_anno_category) AND
    facts3.corpus_ref = facts4.corpus_ref AND
    facts3.left_token = facts4.left_token AND
    facts3.n_na_sample IS TRUE AND
    facts3.node_annotext LIKE 'pos:APPR' AND
    facts3.right_token ^=^ facts4.right_token AND
    facts3.text_ref = facts4.text_ref AND
    facts3.toplevel_corpus IN (442) AND
    facts4.corpus_ref = facts5.corpus_ref AND
    facts4.n_na_sample IS TRUE AND
    facts4.node_annotext LIKE 'lemma\_wsd:tô¹' AND
    facts4.right_token = facts5.left_token - 1 AND
    facts4.text_ref = facts5.text_ref AND
    facts4.toplevel_corpus IN (442) AND
    facts5.corpus_ref = facts6.corpus_ref AND
    facts5.left_token = facts6.left_token AND
    facts5.n_na_sample IS TRUE AND
    facts5.node_annotext LIKE 'pos:NA' AND
    facts5.right_token ^=^ facts6.right_token AND
    facts5.text_ref = facts6.text_ref AND
    facts5.toplevel_corpus IN (442) AND
    facts6.n_na_sample IS TRUE AND
    facts6.node_annotext LIKE 'posLemma:VVINF' AND
    facts6.toplevel_corpus IN (442)

) AS solutions
  LEFT JOIN facts_442 AS v1 ON (v1.toplevel_corpus IN (442) AND v1.toplevel_corpus = solutions.toplevel_corpus AND v1.id = solutions.idvb AND v1.node_annotext LIKE 'pos:%' AND v1.n_na_sample IS TRUE)
  LEFT JOIN facts_442 AS v2 ON (v2.toplevel_corpus IN (442) AND v2.toplevel_corpus = solutions.toplevel_corpus AND v2.id = solutions.idzu AND v2.node_annotext LIKE 'pos:%' AND v2.n_na_sample IS TRUE)

GROUP BY value1, value2

```

**Comment by fab-bar** [comment]
CSV export now works as well.

**Comment by thomaskrause** [comment]
I fixed the frequency query part with #587 by using the user-supplied attributes as basis for ordering the results internally, but I think the CSV exporter is much more tricky to get right.
 
It also poses the question about the semantics of the CSV results, since currently it corresponds to the match definition, which is defined by the order of the attributes in the query. Having this match definition makes sense, as it is defined also for all border cases. Other definitions will struggle with NULL values for attributes present in one disjunction but not others. We also can't change this definition without breaking backwards-compatibility of the REST-API for the `find` function (http://korpling.github.io/ANNIS/doc/interfaceannis_1_1service_1_1_query_service.html#a7aded983819fdc6f0cc143cdded8918a).

As long as the CSV results are just a representation of the match results I don't see how to fix this except as hinting the users to keep in mind the order as a hint in the documentation.
We might be able to fix the CSVMultiTokExporter if we allow a custom parameter that selects the order.

@fab-bar is it ok to close this issue for you with the frequency query fix included?

**Comment by thomaskrause** [comment]
> Just a suggestion for a possible enhancement: maybe problematic queries can be identified and the status box could say something like "Valid query, order differs in disjuncts...". It would be even nicer to have an option to resolve the order automatically by sorting named nodes alphabetically and to the beginning of the query.

Good idea! I added a new issues #588  for that.
